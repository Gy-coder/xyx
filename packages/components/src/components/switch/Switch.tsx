import { forwardRef, ForwardRefRenderFunction, MouseEventHandler } from "react";
import classnames from "classnames";
import { SwitchProps } from "./interface";
import useMergeState from "../../hooks/useMergeState";
import "./index.scss";

const Switch: ForwardRefRenderFunction<any, SwitchProps> = (props, ref) => {
  const {
    defaultChecked,
    checked,
    onChange,
    style,
    className,
    disabled = false,
    size = "middle",
  } = props;
  const [isChecked, setIsChecked, isControlled] = useMergeState(false, {
    defaultValue: defaultChecked,
    value: checked,
  });
  const handleClick: MouseEventHandler<HTMLDivElement> = () => {
    if (disabled) return;
    if (!isControlled) setIsChecked(!isChecked);
    onChange?.(!isChecked);
  };
  return (
    <div
      className={classnames("g-switch", className, {
        [`g-switch-active`]: isChecked,
        [`g-switch-disabled`]: disabled,
        [`g-switch-${size}`]: size,
      })}
      ref={ref}
      style={style}
      onClick={handleClick}
    >
      <div className="g-switch-toggle-wrapper">
        <div className="g-switch-toggle" />
      </div>
    </div>
  );
};

export default forwardRef(Switch);
