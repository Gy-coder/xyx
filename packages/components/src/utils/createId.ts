let id = 0;

export default function createId(){
    return id++
}