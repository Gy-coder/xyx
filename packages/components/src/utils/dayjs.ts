type DayUnit = "year"
    | "month"
    | "day"
    | "hour"
    | "minute"
    | "second"
    | "millisecond";

export default class Dayjs {
    public date: Date

    constructor(date?: number | string | Date) {
        this.date = date ? new Date(date) : new Date()
    }

    get year() {
        return this.date.getFullYear()
    }

    get month() {
        return this.date.getMonth() + 1
    }

    get week() {
        return this.date.getDay()
    }

    get day() {
        return this.date.getDate()
    }

    get hour() {
        return this.date.getHours()
    }

    get minute() {
        return this.date.getMinutes()
    }

    get second() {
        return this.date.getSeconds()
    }

    get millisecond() {
        return this.date.getMilliseconds()
    }

    get raw() {
        return this.date
    }

    public firstDayOfMonth() {
        return new Dayjs(
            new Date(
                this.year,
                this.month - 1,
                1,
                0,
                0,
                0)
        )
    }

    public add(amount: number, unit: DayUnit) {
        let date = new Date(this.date.getTime());
        switch (unit) {
            case 'year':
                const currentDate = date.getDate()
                date.setDate(1)
                date.setFullYear(date.getFullYear() + amount)
                const targetDate = new Date(
                    date.getFullYear(),
                    date.getMonth() + 1,
                    0,
                    0,
                    0,
                    0,
                ).getDate()
                date.setDate(Math.min(currentDate, targetDate))
                break;
            case 'month':
                const d = date.getDate()
                date.setDate(1)
                date.setMonth(date.getMonth() + amount);
                const d2 = new Date(date.getFullYear(), date.getMonth() + 1, 0, 0, 0, 0).getDate()
                date.setDate(Math.min(d, d2))
                break;
            case 'day':
                date.setDate(date.getDate() + amount);
                break;
            case 'hour':
                date.setHours(date.getHours() + amount);
                break;
            case 'minute':
                date.setMinutes(date.getMinutes() + amount);
                break;
            case 'second':
                date.setSeconds(date.getSeconds() + amount);
                break;
            case 'millisecond':
                date.setMilliseconds(date.getMilliseconds() + amount);
                break;

            default:
                throw new Error('Time.add: unknown unit');
        }
        return new Dayjs(date)
    }

    public set(amount: number,unit: DayUnit){
        const date = new Date(this.date.getTime())
        switch (unit){
            case "year":
                date.setFullYear(amount)
                break
            case "month":
                date.setMonth(amount - 1)
                break
            case "day":
                date.setDate(amount)
                break
            case "hour":
                date.setHours(amount)
                break
            case "second":
                date.setSeconds(amount)
                break
            case "millisecond":
                date.setMilliseconds(amount)
                break
            default:
                throw new Error("Time.add: unknown unit")
        }
        return new Dayjs(date)
    }

    isSameDay(day: Dayjs){
        return day.year === this.year && day.month === this.month && day.day === this.day
    }

    isSameMonth(day: Dayjs){
        return day.year === this.year && day.month === this.month
    }

    //@ts-ignore
    format(pattern?: string = 'YYYY-MM-DD') {
        // 目前支持的格式有 YYYY MM DD HH mm ss SSS
        return pattern.replace(/YYYY/g, this.year.toString())
            .replace(/MM/, this.month.toString().padStart(2, '0'))
            .replace(/DD/, this.day.toString().padStart(2, '0'))
            .replace(/HH/, this.hour.toString().padStart(2, '0'))
            .replace(/mm/, this.minute.toString().padStart(2, '0'))
            .replace(/ss/, this.second.toString().padStart(2, '0'))
            .replace(/SSS/, this.millisecond.toString().padStart(3, '0'))
    }
}
