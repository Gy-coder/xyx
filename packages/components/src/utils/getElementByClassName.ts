export function getElementByClassName(container = document, classname: string) {
  return container.getElementsByClassName(classname)[0];
}
