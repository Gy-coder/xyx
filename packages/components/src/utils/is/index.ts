export { default as isUndefined } from "./isUndefined";
export { default as isFunction } from "./isFunction";
