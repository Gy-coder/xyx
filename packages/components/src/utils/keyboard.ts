export const keyboard = {
  Enter: "Enter",
};
