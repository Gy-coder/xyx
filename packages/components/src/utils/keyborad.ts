export const keyborad = {
  Enter: "Enter",
};
