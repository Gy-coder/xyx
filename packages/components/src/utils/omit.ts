export default function omit<T = any>(obj: Record<string, T>, paths: string[]) {
  const res: Record<string, T> = { ...obj },
    set = new Set(paths);
  for (let key in obj) {
    if (set.has(key)) delete res[key];
  }
  return res;
}
