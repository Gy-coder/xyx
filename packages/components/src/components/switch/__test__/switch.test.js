import { screen, render, fireEvent } from "@testing-library/react";
import "@testing-library/jest-dom";
import Switch from "../Switch";
import { getElementByClassName } from "../../../utils/getElementByClassName";
import { useState } from "react";

describe("test Switch component", () => {
  it("should render correctly", () => {
    const { container, asFragment } = render(<Switch />);
    const el = getElementByClassName(container, "g-switch");
    expect(el).toBeVisible();
    expect(asFragment()).toMatchSnapshot();
  });
  it("can change value", () => {
    const jestFn = jest.fn();
    let v, setV, handleChange;
    const Demo = () => {
      [v, setV] = useState(false);
      handleChange = (newV) => {
        setV(newV);
        jestFn(newV);
      };
      return <Switch checked={v} onChange={handleChange} />;
    };
    const { container, asFragment } = render(<Demo />);
    const el = getElementByClassName(container, "g-switch");
    expect(v).toBe(false);
    fireEvent.click(el);
    expect(v).toBe(true);
    expect(jestFn).toBeCalledTimes(1);
    expect(jestFn).toBeCalledWith(true);
    expect(asFragment()).toMatchSnapshot();
  });
  it("disabled", () => {
    const jestFn = jest.fn();
    const { contaier } = render(<Switch disabled onChange={jestFn} />);
    const el = getElementByClassName(contaier, "g-switch");
    expect(el).toHaveClass("g-switch-disabled");
    expect(jestFn).not.toBeCalled();
  });
});
