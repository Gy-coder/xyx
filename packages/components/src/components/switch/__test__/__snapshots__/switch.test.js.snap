// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test Switch component can change value 1`] = `
<DocumentFragment>
  <div
    class="g-switch g-switch-active g-switch-middle"
  >
    <div
      class="g-switch-toggle-wrapper"
    >
      <div
        class="g-switch-toggle"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`test Switch component should render correctly 1`] = `
<DocumentFragment>
  <div
    class="g-switch g-switch-middle"
  >
    <div
      class="g-switch-toggle-wrapper"
    >
      <div
        class="g-switch-toggle"
      />
    </div>
  </div>
</DocumentFragment>
`;
